zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,498 @@
+    ) -> Self {
+        let q_enable = meta.selector();
+        let word_begin = meta.advice_column();
+        let r_multi = meta.advice_column();
+        let input = meta.advice_column();
+        let data = meta.advice_column();
+        let data_rlc = meta.advice_column();
+        let num_bits = meta.advice_column();
+        let shift_factor = meta.advice_column();
+        let decode_factor = meta.advice_column();
+        let huffman_table = array_init::array_init(|_| meta.lookup_table_column());
+
+        meta.enable_equality(data_rlc);
+
+        let mut parts = vec![];
+        meta.create_gate("split parts", |meta| {
+            // next 4 rows as 4 parts
+            for i in 0..MAX_PARTS as i32 {
+                let data = meta.query_advice(data, Rotation(i));
+                let num_bits = meta.query_advice(num_bits, Rotation(i));
+                let r_multi = meta.query_advice(r_multi, Rotation(i));
+                let shift_factor = meta.query_advice(shift_factor, Rotation(i));
+                let decode_factor = meta.query_advice(decode_factor, Rotation(i));
+
+                parts.push(Part {
+                    data,
+                    num_bits,
+                    r_multi,
+                    shift_factor,
+                    decode_factor,
+                });
+            }
+            vec![0.expr()]
+        });
+
+        meta.lookup("huffman table", |meta| {
+            // lookups:
+            //  1. input = huffman_value
+            //  2. data0 * decode_factor0 + data1 * decode_factor1 + data2 * decode_factor2
+            //      + data3 * decode_factor3 = huffman_code
+            //  3. num_bits0 + num_bits1 + num_bits2 + num_bits3 = huffman_code_length
+            let input = meta.query_advice(input, Rotation::cur());
+            let code = decode::data_expr(&parts);
+            let word_begin = meta.query_advice(word_begin, Rotation::cur());
+            vec![
+                (word_begin.clone() * input, huffman_table[0]),
+                (word_begin.clone() * code, huffman_table[1]),
+                (word_begin * decode::num_bits_expr(&parts), huffman_table[2]),
+            ]
+        });
+
+        meta.create_gate("do RLC over parts", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            cb.condition(meta.query_advice(word_begin, Rotation::cur()), |cb| {
+                let mut data_rlc_prev = meta.query_advice(data_rlc, Rotation::prev());
+                for part in parts.iter() {
+                    // 1. TODO: range check(lookup):
+                    //      part.data(0..2^8)
+                    //      part.num_bits(0..32)
+                    //      part.r_multi(bool)
+                    //      part.shift_factor(0..2^32) 2.pow(0..32)
+                    //      part.decode_factor(0..2^32) 2.pow(0..32)
+                    cb.require_boolean("r_multi", part.r_multi.clone());
```

Comment:
If you add the lookup as you described in the comments, `r_multi` will be checked by the lookup so you won't need to check it's boolean using a custom gate, the lookup table will only contain `0` or `1` for the values so the check is already done by the lookup.